File: src.lib/field/kit/particle/Flock.scala
/*                                                                            *\
**           _____  __  _____  __     ____     FieldKit                       **
**          / ___/ / / /____/ / /    /    \    (c) 2009, field                **
**         / ___/ /_/ /____/ / /__  /  /  /    http://www.field.io            **
**        /_/        /____/ /____/ /_____/                                    **
\*                                                                            */
/* created March 31, 2009 */
package field.kit.particle

import field.kit.Logger
import scala.reflect.Manifest

/** 
 * represents a group of particles within the system
 * @author Marcus Wendt
 */
class Flock[P <: Particle](val ps:ParticleSystem)(implicit m:Manifest[P]) 
extends Logger with Collection[P] {
  import scala.collection.mutable.ArrayBuffer
  fine("init")
  
  var behaviours = new ArrayBuffer[Behaviour]
  var emitter = new Emitter[P](this)
  var particles = new ArrayBuffer[P]

  emitter.position(ps.space.center)
  
  def update(dt:Float) = {
    // update emitter / creates new particles
    emitter.update(dt)
      
    // prepare behaviours
    behaviours foreach { b => 
      if(b.isEnabled) b.prepare(dt)
    }

    particles foreach { p =>
      // apply behaviours      
      behaviours foreach { b =>
        if(b.isEnabled) b.apply(p,dt)
      }

      // update particles
      p.update(dt)
      
      // remove dead particles
      if(p.age > p.lifeTime) this -= p
    }
  }
  
  // ---------------------------------------------------------------------------
  // HELPERS
  // ---------------------------------------------------------------------------
  def +=(b:Behaviour) = {
    fine("adding", b)
    b.flock = this
    b.ps = ps
    b.init
    behaviours += b
  }
   
  def +=(p:P) = {
    fine("adding", p)
    p.flock = this
    p.ps = ps
    particles += p
  }
  
  def -=(p:P) = {
    fine("removing", p)
    particles -= p 
  }
  
  def size = particles.size
  
  def elements = particles.elements
}

File: FieldKit.js/src/particle/Flock.js
/*!
 * FieldKit JavaScript Library
 * http://code.google.com/p/fieldkit
 *
 * Copyright (c) 2009 Marcus Wendt
 * Licensed under the LGPL license.
 * http://www.gnu.org/licenses/lgpl.html
 *
 * Created: January 15, 2010
 */

fk.particle.Flock = fk.Class.extend({

	init: function() {
		this.particles = new Array();
		this.behaviours = new Array();
		this.emitter = new fk.particle.Emitter(this);
	},
	
	updateLogic: function(dt) {
		// update emitter
		this.emitter.updateLogic(dt);

		// prepare behaviours
		for(var i=0; i<this.behaviours.length; i++)
			this.behaviours[i].prepare(dt);

		// update particles
		for(var i=0; i<this.particles.length; i++) {
			var p = this.particles[i];

			for(var j=0; j<this.behaviours.length; j++)
				this.behaviours[j].apply(p, dt);

			p.updateLogic(dt);
		}
	},
	
	updateSimulation: function(dt) {
		for(var i=0; i<this.particles.length; i++)
			this.particles[i].updateSimulation(dt);
	},

	add: function(behaviour) { this.behaviours.push(behaviour); },

	toString: function() { return 'Flock'; },
});